File: services/inference_worker/inference_worker/stages/detector.py
"""Stage 1 -- YOLO-based vehicle detection.

Runs a YOLO model (via ONNX Runtime) on full frames to produce bounding boxes
with confidence scores.  The detector outputs a single 'vehicle' class;
fine-grained 12-class discrimination is deferred to Stage 3.
"""

from __future__ import annotations

import logging
from typing import TYPE_CHECKING, Protocol, runtime_checkable

import numpy as np

from inference_worker.models import Detection
from shared_contracts.geometry import BoundingBox

if TYPE_CHECKING:
    from inference_worker.settings import DetectorSettings

logger = logging.getLogger(__name__)

NUM_CLASSES = 12


def _letterbox(
    image: np.ndarray,
    target_size: int,
) -> tuple[np.ndarray, float, tuple[int, int]]:
    """Resize with letterboxing to preserve aspect ratio.

    Returns (padded_image, scale_factor, (pad_w, pad_h)).
    """
    h, w = image.shape[:2]
    scale = target_size / max(h, w)
    new_w, new_h = int(w * scale), int(h * scale)

    resized = np.zeros((target_size, target_size, 3), dtype=np.uint8) + 114
    pad_w = (target_size - new_w) // 2
    pad_h = (target_size - new_h) // 2

    from numpy import ascontiguousarray

    if new_w != w or new_h != h:
        try:
            import cv2

            resized_img = cv2.resize(image, (new_w, new_h), interpolation=cv2.INTER_LINEAR)
        except ImportError:
            from PIL import Image

            pil_img = Image.fromarray(image).resize((new_w, new_h), Image.BILINEAR)
            resized_img = np.array(pil_img)
    else:
        resized_img = image

    resized[pad_h : pad_h + new_h, pad_w : pad_w + new_w] = resized_img
    return ascontiguousarray(resized), scale, (pad_w, pad_h)


def _nms(
    boxes: np.ndarray,
    scores: np.ndarray,
    iou_threshold: float,
) -> list[int]:
    """Non-maximum suppression on (N, 4) boxes in xyxy format."""
    if len(boxes) == 0:
        return []

    x1, y1, x2, y2 = boxes[:, 0], boxes[:, 1], boxes[:, 2], boxes[:, 3]
    areas = (x2 - x1) * (y2 - y1)
    order = scores.argsort()[::-1]
    keep: list[int] = []

    while order.size > 0:
        i = order[0]
        keep.append(int(i))
        if order.size == 1:
            break

        xx1 = np.maximum(x1[i], x1[order[1:]])
        yy1 = np.maximum(y1[i], y1[order[1:]])
        xx2 = np.minimum(x2[i], x2[order[1:]])
        yy2 = np.minimum(y2[i], y2[order[1:]])

        inter = np.maximum(0.0, xx2 - xx1) * np.maximum(0.0, yy2 - yy1)
        iou = inter / (areas[i] + areas[order[1:]] - inter + 1e-6)

        remaining = np.where(iou <= iou_threshold)[0]
        order = order[remaining + 1]

    return keep


@runtime_checkable
class DetectorBackend(Protocol):
    """Protocol for swappable detection backends (ONNX, TorchScript, stub)."""

    def detect(
        self,
        preprocessed: np.ndarray,
    ) -> np.ndarray:
        """Run detection on a preprocessed (1, 3, H, W) float32 tensor.

        Returns raw model output whose shape depends on the backend.
        """
        ...


class OnnxDetectorBackend:
    """ONNX Runtime detection backend."""

    def __init__(self, model_path: str) -> None:
        import onnxruntime as ort

        providers = ["CUDAExecutionProvider", "CPUExecutionProvider"]
        self._session = ort.InferenceSession(model_path, providers=providers)
        self._input_name = self._session.get_inputs()[0].name
        logger.info("Loaded ONNX detector from %s", model_path)

    def detect(self, preprocessed: np.ndarray) -> np.ndarray:
        outputs = self._session.run(None, {self._input_name: preprocessed})
        return outputs[0]


class StubDetectorBackend:
    """Deterministic stub for testing without a real model file."""

    def detect(self, preprocessed: np.ndarray) -> np.ndarray:
        return np.empty((1, 0, 6), dtype=np.float32)


class VehicleDetector:
    """Stage 1: full-frame vehicle detection with YOLO.

    Accepts raw BGR/RGB frames, applies letterbox preprocessing, runs the
    detection model, and returns normalised BoundingBox detections.
    """

    def __init__(self, settings: DetectorSettings, backend: DetectorBackend | None = None) -> None:
        self._settings = settings
        if backend is not None:
            self._backend = backend
        else:
            try:
                self._backend = OnnxDetectorBackend(settings.model_path)
            except Exception:
                logger.warning("Could not load ONNX detector; falling back to stub")
                self._backend = StubDetectorBackend()

    def detect_frame(
        self,
        frame: np.ndarray,
        frame_index: int,
    ) -> list[Detection]:
        """Run detection on a single frame (H, W, 3) uint8 array.

        Returns a list of Detection objects with normalised bbox coordinates.
        """
        h_orig, w_orig = frame.shape[:2]
        target = self._settings.input_size

        padded, scale, (pad_w, pad_h) = _letterbox(frame, target)

        blob = padded.astype(np.float32) / 255.0
        blob = blob.transpose(2, 0, 1)[np.newaxis]  # (1, 3, H, W)

        raw = self._backend.detect(blob)
        return self._postprocess(raw, w_orig, h_orig, scale, pad_w, pad_h, frame_index)

    def _postprocess(
        self,
        raw: np.ndarray,
        w_orig: int,
        h_orig: int,
        scale: float,
        pad_w: int,
        pad_h: int,
        frame_index: int,
    ) -> list[Detection]:
        """Convert raw model output to normalised Detection objects.

        Handles the common YOLO output format: (1, N, 4+1+num_classes) or
        (1, 4+1+num_classes, N) where the 4 values are cx, cy, w, h.
        """
        if raw.size == 0:
            return []

        if raw.ndim == 3:
            preds = raw[0]
        else:
            preds = raw

        if preds.ndim == 2 and preds.shape[0] in (4, 5, 5 + NUM_CLASSES, 4 + NUM_CLASSES):
            if preds.shape[0] < preds.shape[1]:
                preds = preds.T

        if preds.shape[1] < 5:
            return []

        conf_threshold = self._settings.confidence_threshold
        target = self._settings.input_size

        if preds.shape[1] == 5:
            scores = preds[:, 4]
        elif preds.shape[1] == 5 + NUM_CLASSES:
            scores = preds[:, 4] * preds[:, 5:].max(axis=1)
        elif preds.shape[1] == 4 + NUM_CLASSES:
            scores = preds[:, 4:].max(axis=1)
        else:
            scores = preds[:, 4]

        mask = scores >= conf_threshold
        preds = preds[mask]
        scores = scores[mask]

        if len(preds) == 0:
            return []

        cx, cy, bw, bh = preds[:, 0], preds[:, 1], preds[:, 2], preds[:, 3]
        x1 = cx - bw / 2
        y1 = cy - bh / 2
        x2 = cx + bw / 2
        y2 = cy + bh / 2

        boxes_xyxy = np.stack([x1, y1, x2, y2], axis=1)

        keep = _nms(boxes_xyxy, scores, self._settings.nms_iou_threshold)
        keep = keep[: self._settings.max_detections]

        detections: list[Detection] = []
        for idx in keep:
            bx1 = (boxes_xyxy[idx, 0] - pad_w) / scale
            by1 = (boxes_xyxy[idx, 1] - pad_h) / scale
            bx2 = (boxes_xyxy[idx, 2] - pad_w) / scale
            by2 = (boxes_xyxy[idx, 3] - pad_h) / scale

            nx = max(0.0, min(1.0, bx1 / w_orig))
            ny = max(0.0, min(1.0, by1 / h_orig))
            nw = max(0.0, min(1.0, (bx2 - bx1) / w_orig))
            nh = max(0.0, min(1.0, (by2 - by1) / h_orig))

            if nw < 0.005 or nh < 0.005:
                continue

            detections.append(
                Detection(
                    bbox=BoundingBox(x=nx, y=ny, w=nw, h=nh),
                    confidence=float(scores[idx]),
                    frame_index=frame_index,
                )
            )

        return detections

File: apps/mobile_flutter/lib/core/inference/detector.dart
/// Stage 1 — YOLO-based vehicle detection via TFLite.
///
/// Runs a YOLOv8 model on full frames to produce bounding boxes with
/// confidence scores. The detector outputs a single "vehicle" class;
/// fine-grained 12-class discrimination is deferred to Stage 3.
///
/// Ported from `services/inference_worker/inference_worker/stages/detector.py`.
library;

import 'dart:math' as math;
import 'dart:typed_data';

import 'package:greyeye_mobile/core/inference/models.dart';
import 'package:greyeye_mobile/core/inference/pipeline_settings.dart';
import 'package:image/image.dart' as img;
import 'package:tflite_flutter/tflite_flutter.dart';

class VehicleDetector {
  VehicleDetector(this._settings);

  final DetectorSettings _settings;
  Interpreter? _interpreter;

  /// Number of classes the loaded model outputs.
  /// Inferred from the output tensor shape during [load].
  int _numClasses = 0;

  int get numClasses => _numClasses;

  Future<void> load() async {
    final interpreter = await Interpreter.fromAsset(_settings.modelPath);
    _interpreter = interpreter;

    final outputShape = interpreter.getOutputTensor(0).shape;
    _numClasses = _inferNumClasses(outputShape);
  }

  /// Load from raw model bytes (for use in background isolates where
  /// Flutter asset bindings are unavailable).
  void loadFromBuffer(Uint8List bytes) {
    final interpreter = Interpreter.fromBuffer(bytes);
    _interpreter = interpreter;

    final outputShape = interpreter.getOutputTensor(0).shape;
    _numClasses = _inferNumClasses(outputShape);
  }

  /// Derive the class count from the model's output tensor shape.
  ///
  /// YOLOv8 transposed `[1, 4+C, N]` (rows < cols) -> `C = predDim - 4`.
  /// YOLOv5 standard  `[1, N, 5+C]` (rows > cols)  -> `C = predDim - 5`.
  /// Single-class     `[1, N, 5]`                   -> `C = 1`.
  static int _inferNumClasses(List<int> shape) {
    if (shape.length == 3) {
      final rows = shape[1];
      final cols = shape[2];
      if (rows < cols) {
        // Transposed (YOLOv8): predDim = rows = 4 + numClasses
        return rows > 4 ? rows - 4 : 0;
      } else {
        // Transposed by TFLite export (NHWC layout): predDim = cols = 4 + numClasses
        if (cols == 5) return 1;
        return cols > 4 ? cols - 4 : 0;
      }
    } else if (shape.length == 2) {
      final predDim = shape[1];
      if (predDim == 5) return 1;
      return predDim > 4 ? predDim - 4 : 0;
    }
    return 0;
  }

  void dispose() {
    _interpreter?.close();
    _interpreter = null;
  }

  /// Run detection on a single frame.
  ///
  /// [frame] is a decoded image. Returns normalised [Detection] objects.
  List<Detection> detectFrame(img.Image frame, int frameIndex) {
    final interpreter = _interpreter;
    if (interpreter == null) return [];

    final target = _settings.inputSize;
    final hOrig = frame.height;
    final wOrig = frame.width;

    // --- Letterbox ---
    final scale = target / math.max(hOrig, wOrig);
    final newW = (wOrig * scale).round();
    final newH = (hOrig * scale).round();
    final padW = (target - newW) ~/ 2;
    final padH = (target - newH) ~/ 2;

    final resized = img.copyResize(frame, width: newW, height: newH);

    // Build [1, target, target, 3] float32 input with 114/255 padding.
    final inputList = List.generate(
      1,
      (_) => List.generate(
        target,
        (y) => List.generate(
          target,
          (x) {
            final srcY = y - padH;
            final srcX = x - padW;
            if (srcX >= 0 && srcX < newW && srcY >= 0 && srcY < newH) {
              final pixel = resized.getPixel(srcX, srcY);
              return [
                pixel.r / 255.0,
                pixel.g / 255.0,
                pixel.b / 255.0,
              ];
            }
            return [114.0 / 255.0, 114.0 / 255.0, 114.0 / 255.0];
          },
        ),
      ),
    );

    // Allocate output buffer matching the model's output tensor shape.
    final outputShape = interpreter.getOutputTensor(0).shape;
    final output = _allocateOutput(outputShape);

    interpreter.run(inputList, output);

    // Flatten the output to a Float32List for postprocessing.
    final flatOutput = _flattenOutput(output);

    return _postprocess(
      flatOutput,
      outputShape,
      wOrig,
      hOrig,
      scale,
      padW,
      padH,
      frameIndex,
    );
  }

  List<Detection> _postprocess(
    Float32List raw,
    List<int> shape,
    int wOrig,
    int hOrig,
    double scale,
    int padW,
    int padH,
    int frameIndex,
  ) {
    if (raw.isEmpty) return [];

    // Determine prediction matrix dimensions.
    late int numPreds;
    late int predDim;
    late bool transposed;

    if (shape.length == 3) {
      final rows = shape[1];
      final cols = shape[2];
      if (rows < cols) {
        // (1, features, N) — transposed layout
        numPreds = cols;
        predDim = rows;
        transposed = true;
      } else {
        numPreds = rows;
        predDim = cols;
        transposed = false;
      }
    } else if (shape.length == 2) {
      numPreds = shape[0];
      predDim = shape[1];
      transposed = false;
    } else {
      return [];
    }

    if (_numClasses == 0 ||
        (predDim != 5 &&
            predDim != 5 + _numClasses &&
            predDim != 4 + _numClasses)) {
      return [];
    }

    // Detect whether model outputs normalized (0-1) or pixel-space coords.
    // Stage1/Stage2 models converted from CoreML output normalized coords;
    // legacy detector.tflite outputs pixel-space coords.
    final target = _settings.inputSize.toDouble();
    final isNormalized = _isNormalizedOutput(raw, numPreds, predDim, transposed);
    final coordScale = isNormalized ? target : 1.0;

    // Extract predictions into a 2D view.
    final confThreshold = _settings.confidenceThreshold;
    final filteredBoxes = <List<double>>[];
    final filteredScores = <double>[];
    final filteredClassIndices = <int>[];

    for (var n = 0; n < numPreds; n++) {
      final p = Float64List(predDim);
      for (var f = 0; f < predDim; f++) {
        p[f] = transposed ? raw[f * numPreds + n] : raw[n * predDim + f];
      }

      double score;
      int bestClassIdx = 0;
      if (predDim == 5) {
        score = p[4];
      } else if (predDim == 5 + _numClasses) {
        var maxCls = p[5];
        bestClassIdx = 0;
        for (var c = 6; c < predDim; c++) {
          if (p[c] > maxCls) {
            maxCls = p[c];
            bestClassIdx = c - 5;
          }
        }
        score = p[4] * maxCls;
      } else if (predDim == 4 + _numClasses) {
        var maxCls = p[4];
        bestClassIdx = 0;
        for (var c = 5; c < predDim; c++) {
          if (p[c] > maxCls) {
            maxCls = p[c];
            bestClassIdx = c - 4;
          }
        }
        score = maxCls;
      } else {
        score = p[4];
      }

      if (score < confThreshold) continue;

      if (_settings.filterVehiclesOnly &&
          !_settings.cocoClassMap.containsKey(bestClassIdx)) {
        continue;
      }

      final cx = p[0] * coordScale;
      final cy = p[1] * coordScale;
      final bw = p[2] * coordScale;
      final bh = p[3] * coordScale;
      filteredBoxes.add([cx - bw / 2, cy - bh / 2, cx + bw / 2, cy + bh / 2]);
      filteredScores.add(score);
      filteredClassIndices.add(bestClassIdx);
    }

    if (filteredBoxes.isEmpty) return [];

    // NMS
    final keepIndices = _nms(
      filteredBoxes,
      filteredScores,
      _settings.nmsIouThreshold,
    );
    final maxDet = _settings.maxDetections;

    final detections = <Detection>[];
    for (var ki = 0; ki < keepIndices.length && ki < maxDet; ki++) {
      final idx = keepIndices[ki];
      final box = filteredBoxes[idx];

      final bx1 = (box[0] - padW) / scale;
      final by1 = (box[1] - padH) / scale;
      final bx2 = (box[2] - padW) / scale;
      final by2 = (box[3] - padH) / scale;

      final nx = (bx1 / wOrig).clamp(0.0, 1.0);
      final ny = (by1 / hOrig).clamp(0.0, 1.0);
      final nw = ((bx2 - bx1) / wOrig).clamp(0.0, 1.0);
      final nh = ((by2 - by1) / hOrig).clamp(0.0, 1.0);

      if (nw < 0.005 || nh < 0.005) continue;

      final rawClassIdx = filteredClassIndices[idx];
      final mappedClassCode = _settings.cocoClassMap.containsKey(rawClassIdx)
          ? _settings.cocoClassMap[rawClassIdx]
          : (rawClassIdx + 1);

      detections.add(
        Detection(
          bbox: BoundingBox(x: nx, y: ny, w: nw, h: nh),
          confidence: filteredScores[idx],
          frameIndex: frameIndex,
          classCode: mappedClassCode,
        ),
      );
    }

    return detections;
  }

  /// Check whether the model outputs normalized (0-1) bbox coordinates by
  /// sampling the first few predictions. If the max bbox coordinate value
  /// across sampled predictions is <= 1.0, the output is normalized.
  static bool _isNormalizedOutput(
    Float32List raw,
    int numPreds,
    int predDim,
    bool transposed,
  ) {
    var maxCoord = 0.0;
    final samplesToCheck = math.min(numPreds, 100);
    for (var n = 0; n < samplesToCheck; n++) {
      for (var f = 0; f < 4; f++) {
        final val = transposed ? raw[f * numPreds + n] : raw[n * predDim + f];
        if (val.abs() > maxCoord) maxCoord = val.abs();
      }
    }
    return maxCoord <= 1.0;
  }
}

/// Allocate a nested List matching [shape] for TFLite output.
Object _allocateOutput(List<int> shape) {
  if (shape.length == 1) {
    return List<double>.filled(shape[0], 0.0);
  }
  return List.generate(shape[0], (_) {
    return _allocateOutput(shape.sublist(1));
  });
}

/// Flatten a nested List output into a [Float32List].
Float32List _flattenOutput(Object nested) {
  final flat = <double>[];
  _flattenRecursive(nested, flat);
  return Float32List.fromList(flat);
}

void _flattenRecursive(Object obj, List<double> out) {
  if (obj is List) {
    for (final item in obj) {
      _flattenRecursive(item as Object, out);
    }
  } else if (obj is double) {
    out.add(obj);
  } else if (obj is num) {
    out.add(obj.toDouble());
  }
}

/// Non-maximum suppression on xyxy boxes.
List<int> _nms(
  List<List<double>> boxes,
  List<double> scores,
  double iouThreshold,
) {
  if (boxes.isEmpty) return [];

  final indices = List<int>.generate(boxes.length, (i) => i);
  indices.sort((a, b) => scores[b].compareTo(scores[a]));

  final keep = <int>[];
  final suppressed = List<bool>.filled(boxes.length, false);

  for (final i in indices) {
    if (suppressed[i]) continue;
    keep.add(i);

    final bi = boxes[i];
    final areaI = (bi[2] - bi[0]) * (bi[3] - bi[1]);

    for (final j in indices) {
      if (suppressed[j] || j == i) continue;

      final bj = boxes[j];
      final xx1 = math.max(bi[0], bj[0]);
      final yy1 = math.max(bi[1], bj[1]);
      final xx2 = math.min(bi[2], bj[2]);
      final yy2 = math.min(bi[3], bj[3]);

      final inter = math.max(0.0, xx2 - xx1) * math.max(0.0, yy2 - yy1);
      final areaJ = (bj[2] - bj[0]) * (bj[3] - bj[1]);
      final iou = inter / (areaI + areaJ - inter + 1e-6);

      if (iou > iouThreshold) {
        suppressed[j] = true;
      }
    }
  }

  return keep;
}
